### PR TITLE
Add conformer generation tool

### PR DESCRIPTION
**Pull request by mattwthompson**
## Description

Much of this was already done in https://github.com/openforcefield/openforcefield/pull/661 before we decided to move it here

* ~~The module structure is temporary and likely to live elsewhere in the future~~ It has a home now!
* I'd like us to think about how to test these utilities more rigorously than we might normally do in a unit test, i.e. add some safeguards that the results are reasonable in addition to ensuring that meet the requirements of the spec

- [x] Tag issue being addressed
- [x] Add tests
- [ ] Update docstrings/documentation
- [ ] Update changelog

### TODO:
- [x] De-duplicate conformers after OpenMM minimization
- [x] Fix `return_topology` issue with partial charges
~~- [ ] Ensure it works with OFFTK 0.7.1 but can take advantage of https://github.com/openforcefield/openforcefield/pull/670 (0.8.0+)~~
- [x] Depend on OFFTK > 0.7.1

### Tests to include

Function calls
- [x] loading one molecule from a .sdf file WITHOUT charges
- [x] loading one molecule from a .sdf file WITH charges
- [x] (OE only) loading one molecule from a .mol2 file WITH charges
- [ ] (OE only) loading one molecule from a .mol2 file WITHOUT charges (is this even possible)
- [x] loading a molecule with a defined name (from any format) and ensuring the output file has that prefix
- [x] loading a molecule with a defined name (from any format), and providing a -p option and ensuring the output file has the -p prefix
- [x] loading a molecule with a defined name (from any format) and ensuring the output file has that prefix
- [ ] loading multiple molecules from a .sdf file
- [x] loading one molecule from SMILES in a .smi file
- [x] loading multiple molecules from SMILES in a .smi file
- [x] loading a molecule with ambiguous stereo from SMILES and enumerating stereoisomers
- [x] loading a molecule with defined stereo from SMILES and preserving that stereochemistry
- [x] run a single molcule through all `get_available_forcefields()`

Direct CLI calls
- [x] Ensure `--prefix` overrides original filename
- [ ] Ensure conformers are enumerated
- [ ] Ensure stereoisomers are enumerated?
- [ ] Ensure SD data tags are accurately written
- [x] Test different forms of `--forcefield`
- [x] Test different forms of `--toolkit`
- [x] Gracefully handle cases  of arguments requesting unavailable resources (force field versions, un-installed/licensed toolkits)

Out of scope here but might be useful down the line:
- [ ] Further integration with `openmmforcefields`
- [ ] Inputting/outputting data to STDOUT
- [ ] Keep some (?) conformers, if found in input files
- [ ] Piping several calls together

### CODE REVIEW

**Comment by mattwthompson** [comment]
@j-wags some discrete questions for clarification:

1.  For tests `loading a molecule with a defined name (from any format) ...` do you mean name like `my_cool_mol.sdf` or what populates the `Molecule.from_file('mol.sdf').name` attribute?
1. Should a multi-molecule SDF file be returned as separate SDF files, one for each molecule?
1. Can we get away with just testing the functions, or should we also have tests that directly call `python  /path/to/scripy.py --args args`? It seems most CLIs only do the former and trust `click`/`argparse`/etc do behave as expected - alone this seems reasonable but I'm getting hung up on how the user experience isn't directly tested.

**Comment by j-wags** [comment]
Ah, these are hard. Thanks for digging into this. 

> For tests loading a molecule with a defined name (from any format) ... do you mean name like my_cool_mol.sdf or what populates the Molecule.from_file('mol.sdf').name attribute?

Let's do file name/prefix. Various file formats may or may not populate `mol.name`, so any behavior dependent on that will be undefined/sporadic.

> Should a multi-molecule SDF file be returned as separate SDF files, one for each molecule?

Yes. I'd also encourage the output conformers to be different files as well, since that will mirror our "SDFs are never mutli-conformer" input behavior as well. 

> Can we get away with just testing the functions, or should we also have tests that directly call python /path/to/scripy.py --args args? It seems most CLIs only do the former and trust click/argparse/etc do behave as expected - alone this seems reasonable but I'm getting hung up on how the user experience isn't directly tested.

I'd still like to have a few direct tests of the UX., but we don't need tons. Maybe a minimal number is:
* One with as few arguments as possible so that we can ensure that any default arguments are hitting their marks
* One with as many arguments as possible, to ensure they're all wired up correctly.

**Comment by mattwthompson** [comment]
@j-wags couple more

* Would it be ok to require OpenFF Toolkit  > 0.7.1?
* I have to admit I don't fully understand how conformer generation is supposed to work; is it supposed to start with `mol.generate_confomers(n_conformers=some_large_number)`, minimizing against Parsley, and then screening out bad conformers? What are the criteria for these steps?
* Should we support non-mainline force fields? In particular, SMIRNOFF99Frosst (hooking in other things like GAFF via `openmmforcefields` would be a nice feature but out of scope here)
* The current setup is exclusively file I/O, i.e. taking in some file and writing out. Should there eventually be an option to instead pipe the contents of the SDF file(s) to STDOUT?

File: openff/cli/generate_conformers.py [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+                for i, iso in enumerate(stereoisomers):
+                    iso.name = mol.name + "_stereoisomer" + str(i)
+                    mols_with_unpacked_stereoisomers.append(iso)
+            else:
+                mols_with_unpacked_stereoisomers.append(mol)
+        mols = mols_with_unpacked_stereoisomers
+
+    # If no conformers were found (i.e. SMI files), generate some
+    # TODO: How many should be generated?
+    # TODO: If 1 or a few conformers are found, should more be generated?
```
Comment (j-wags) [inline openff/cli/generate_conformers.py]: I think the proper behavior here would be to _always_ generate conformers. If we can keep the input conformer(s) as a starting point for minimizations, that's a plus, but not absolutely necessary.

Folks who want to do something with some pre-existing conformers should use our `conformer_energies` tool.

**Comment by j-wags** [comment]
> Would it be ok to require OpenFF Toolkit > 0.7.1?

Yes. Thanks for asking.

> I have to admit I don't fully understand how conformer generation is supposed to work; is it supposed to start with `mol.generate_confomers(n_conformers=some_large_number)`, minimizing against Parsley, and then screening out bad conformers? What are the criteria for these steps?

Yup, you've pretty much got it right. I might not even bother with screening out "bad conformers" in the first edition of this script. Without it, the workflow should look something like (with a lot of handwaving):

```
for mol in dataset:
    # We set the rms_cutoff arg here to a small value, so we get lots of starting conformers
    mol.generate_conformers(n_conformers=some_large_number, rms_cutoff=0.25*unit.angstrom)
    sys = ff.create_openmm_system(mol)
    min_confs = []
    for conf in mol.conformers:
        new_conf = minimize_coords(sys, conf)
        new_energy = calc_energy(sys, new_conf)
        min_confs.append([new_conf, new_energy])
    # Sort from low to high energy, so the minimum energy confs are written first
    min_confs.sort(key=lambda x: x[1])
    # If we want to add logic to remove high-energy conformers in a subsequent release, we'd do the filtering here
    mol_copy = Molecule(mol)
    for idx, (min_conf, energy) in enumerate(min_confs):
        mol_copy.conformers = None
        mol_copy.add_conformer(min_conf)
        mol_copy.properties['minimized_energy_kcal_mol'] = energy / unit.value_in_unit(unit.kilocalorie_per_mole)
        mol_copy.write(f'{mol_name}_conf{idx}.sdf', file_format='sdf')
```
(note the setting of a low RMS cutoff in `generate_conformers`, so we get lots of closely-spaced conformers)
(and I also forgot to add the conditional logic for pre-existing molecule charges,a nd teh grab the final ones from `return_topology`)

> Should we support non-mainline force fields? In particular, SMIRNOFF99Frosst (hooking in other things like GAFF via openmmforcefields would be a nice feature but out of scope here)

Let's try to support everything that we can dig up from `get_available_force_fields`. But we can save openmmforcefields integration/GAFF, for a subsequent release.

> The current setup is exclusively file I/O, i.e. taking in some file and writing out. Should there eventually be an option to instead pipe the contents of the SDF file(s) to STDOUT?

Eventually, yes. It's not clear exactly what format would be intended from the piping (since the file name is important in the context of the output -- It indicates the parent molecule and the energy ranking of this conformer)

**Comment by SimonBoothroyd** [comment]
I'd recommend using [`click`](https://github.com/pallets/click) (probably with [click-option-group](https://github.com/click-contrib/click-option-group)) rather than `argparse`.

Very nice to use, creates nice help pages, and can easily and programatically create new CLI's e.g. [I programmatically create some in `nonbonded`.](https://github.com/SimonBoothroyd/nonbonded/blob/master/nonbonded/cli/projects/plots.py)

File: openff/cli/generate_conformers.py [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+        if len(raw_mols) > 1:
+            mol.name += str(i)
+        mols.append(mol)
+
+    # TODO: How to handle names of different stereoisomers? Just act like they're different conformers?
+    if ambiguous_stereochemistry:
+        mols_with_unpacked_stereoisomers = []
+        for mol in mols:
+            # TODO: This is a brute-force approach, it would be better to check stereo
+            #  without needing to call enumerate_stereoisomers
```
Comment (j-wags) [inline openff/cli/generate_conformers.py:73]: (not blocking) Agreed. Hopefully we can hit this in the molecule refactor, with a separate `stereo_possible` field on atoms and bonds.

File: openff/cli/generate_conformers.py [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+    for i, mol in enumerate(raw_mols):
+        if prefix is not None:
+            mol.name = prefix
+        elif not mol.name:
+            mol.name = "molecule"
+        if len(raw_mols) > 1:
+            mol.name += str(i)
+        mols.append(mol)
+
+    # TODO: How to handle names of different stereoisomers? Just act like they're different conformers?
```
Comment (j-wags) [inline openff/cli/generate_conformers.py:68]: (not blocking) Oof. This is hard. I think what you've implemented currently is good. 

I was initially going to say that we should merge them under the same name, but then they'd ACTUALLY have different parent molecules, so the bookkeeping to keep track of relative conformer energies would go out the window. 

If this becomes a pain point, we could add a command line argument for whether to treat enumerated stereoisomers as separate molecules.

File: openff/cli/generate_conformers.py [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+                mols_with_unpacked_stereoisomers.append(mol)
+        mols = mols_with_unpacked_stereoisomers
+
+    for mol in mols:
+        # TODO: Keep some conformers, if found in input molecule?
+        mol.generate_conformers(
+            toolkit_registry=registry,
+            n_conformers=20,
+            rms_cutoff=0.25 * unit.angstrom,
+        )
```
Comment (j-wags) [inline openff/cli/generate_conformers.py]: (not blocking) I think so. However, generate_conformers will delete existing conformers. To keep them, we could use something like 
```suggestion
    for mol in mols:
        existing_conf = None
        if mol.conformers:
           existing_conf = copy.deepcopy(mol.conformers[0])
        mol.generate_conformers(
            toolkit_registry=registry,
            n_conformers=100,
            rms_cutoff=0.25 * unit.angstrom,
        )
        if existing_conf:
            mol.add_conformers(existing_conf)
```

File: openff/cli/generate_conformers.py [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+
+    # TODO: What happens if some molecules in a multi-molecule file have charges, others don't?
```
Comment (j-wags) [inline openff/cli/generate_conformers.py:95]: (not blocking) Hm, that's a tricky case. I just skimmed the implementation, and what will happen is 
* if ANY instances of this molecule have charges, those will be used for ALL instances
* if MULTIPLE instances of this molecule have charges, the first one will be used to assign charges to ALL instances of the molecule. Charges on subsequent instances of the same graph molecule will be ignored.

(I've made [an Issue](https://github.com/openforcefield/openforcefield/issues/732) to implement a check+error in OFFTK for this. )

~A possible bug in this workflow is that, if an enumerated stereoisomer of a molecule with charges is being parameterized, then its graph may not match. We should check whether enumerated stereoisomers inherit the charges of their parents.~ Phew -- They do!

File: openff/cli/generate_conformers.py [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+    mols_with_charges = []
+    for mol in mols:
+        if mol.partial_charges is not None:
+            mols_with_charges.append(mol)
+
+    mols = _collapse_conformers(mols)
```
Comment (j-wags) [inline openff/cli/generate_conformers.py]: (blocking) -- I think this should be much higher up in the workflow (basically soon after the file read). That would clarify where we make decisions about stuff like the multiple-isomorphic-charge-molecules question above.

File: openff/cli/generate_conformers.py [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+
+    mols_out = []
+    for mol in mols:
+        simulation, partial_charges = _build_simulation(
+            molecule=mol, forcefield=ff, mols_with_charge=mols_with_charges
```
Comment (j-wags) [inline openff/cli/generate_conformers.py]: (non-blocking) Complications like the multiple-isomorphic-charge-molecules problem might be solved by 
```suggestion
            molecule=mol, forcefield=ff, mols_with_charge=[mol]
```

File: openff/cli/generate_conformers.py [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+                collapsed_molecules[-1].add_conformer(conformer)
+        else:
+            collapsed_molecules.append(molecule)
+
+    return collapsed_molecules
+
+
+def _build_simulation(molecule, forcefield, mols_with_charge):
+    """Given a Molecule and ForceField, initialize a barebones OpenMM Simulation."""
+    mol_copy = deepcopy(molecule)
```
Comment (j-wags) [inline openff/cli/generate_conformers.py]: (blocking) Is this ever used?
```suggestion

```

File: openff/cli/tests/test_cli_functions.py [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+
+        from openff.cli import __version__ as cli_version
+
+        assert "OpenFF Toolkit" in out
+        assert "OpenFF CLI" in out
+        assert toolkit_version in out
+        assert cli_version in out
+
+
+# TODO: Update skipifs after #615
```
Comment (j-wags) [inline openff/cli/tests/test_cli_functions.py]: (not blocking) [#615](https://github.com/openforcefield/openforcefield/pull/615) is now merged, though at a quick glance I don't know how it affects this.

File: openff/cli/tests/test_cli_functions.py [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+@pytest.mark.skipif(not RDKIT_AVAILABLE, reason="requires rdkit")
+@pytest.mark.skipif(not OPENEYE_AVAILABLE, reason="requires openeye")
+@pytest.mark.parametrize("toolkit", ["rdkit", "openeye"])
```
Comment (j-wags) [inline openff/cli/tests/test_cli_functions.py]: (blocking) Could we add a simple test that runs on an RDKit-only build? I would just want to make sure that the workflow isn't quietly reliant on some little OE-dependent funcitonality.

**Review by j-wags — CHANGES_REQUESTED** [review]
Great, clear comments. This is an absolute masterpiece. I've taken my time reviewing this and asked for a few things around tricky corner cases, and some rearrangement that might make the state of the molecules at different points a bit clearer. 

Also, I'm seeing how much complexity is added by reading multi-molecule SDFs. Perhaps in the future we could adopt a strict "one molecule, one conformer" philosophy to speed up development.

File: openff/cli/tests/test_cli_functions.py (hunk 9 above)
Comment (mattwthompson) [reply openff/cli/tests/test_cli_functions.py]: It was more or less cosmetic

File: openff/cli/generate_conformers.py (hunk 8 above)
Comment (mattwthompson) [reply openff/cli/generate_conformers.py]: I think it was used at some point, but it looks like it's not used now.

File: openff/cli/generate_conformers.py (hunk 2 above)
Comment (mattwthompson) [reply openff/cli/generate_conformers.py:73]: Opened a feature request to have a stub: https://github.com/openforcefield/openforcefield/issues/739 but resolving this for now

File: openff/cli/generate_conformers.py (hunk 3 above)
Comment (mattwthompson) [reply openff/cli/generate_conformers.py:68]: Keep track of this in #11 but resolving here

File: openff/cli/generate_conformers.py (hunk 5 above)
Comment (mattwthompson) [reply openff/cli/generate_conformers.py:95]: I think this is getting into the category of edge cases that are less worth worrying about in the initial implementation, so I intend to leave this be. Since it's documented, so we can loop back to this later (and we probably want to)

File: openff/cli/tests/test_cli_functions.py (hunk 10 above)
Comment (mattwthompson) [reply openff/cli/tests/test_cli_functions.py]: I've updated the pytest setup so this class's tests actually run  on the RDKit-only builds (after #10). Will check the build logs later but it's working as expected locally

```
======================================== 41 passed, 42 skipped, 1 warning in 136.18s (0:02:16) =========================================

**Comment by mattwthompson** [comment]
@j-wags thanks again for all the feedback. I think I've addressed everything, but the changes are enough to justify another pass through

**Review by j-wags — APPROVED** [review]
Awesome. Thanks, Matt! This looks great.

**Comment by mattwthompson** [comment]
Thanks for the detailed spec and thorough reviews @j-wags!
